gem5+SST13 Changes

kaustav-goswami opened the pull request:
This PR compares the disaggregated memory changes.

kaustav-goswami committed:
ext: Updated the gem5 SST Bridge to use SST 13.0.0

This change updates the gem5 SST Bridge to use SST 13.0.0. Changes
are made to replace SimpleMem class to StandardMem class as
SimpleMem will be deprecated in SST 14 and above. In addition, the
translator.hh is updated to translate more types of gem5 packets.
A new parameter `ports` was added on SST's side when invoking the
gem5 component which does not require recompiling the gem5
component whenever a new outgoing bridge is added in a gem5 config.

Change-Id: I45f0013bc35d088df0aa5a71951422cabab4d7f7
Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext: Updated the gem5 SST Bridge to use SST 13.0.0

This change updates the gem5 SST Bridge to use SST 13.0.0. Changes
are made to replace SimpleMem class to StandardMem class as
SimpleMem will be deprecated in SST 14 and above. In addition, the
translator.hh is updated to translate more types of gem5 packets.
A new parameter `ports` was added on SST's side when invoking the
gem5 component which does not require recompiling the gem5
component whenever a new outgoing bridge is added in a gem5 config.

Change-Id: I45f0013bc35d088df0aa5a71951422cabab4d7f7
Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext,configs: adding a disaggregated memory setup

This change adds necessasry tools to simulate multiple gem5
nodes to simulate a disaggregated memory setup using gem5 and SST.

Change-Id: I6e1e52d4ba8df7c161b3151c9b2c02b72fc7cc31
Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext,configs: adding a disaggregated memory setup

This change adds necessasry tools to simulate multiple gem5
nodes to simulate a disaggregated memory setup using gem5 and SST.

Change-Id: I6e1e52d4ba8df7c161b3151c9b2c02b72fc7cc31
Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext: added functional accesses to gem5/SST bridge.

This change adds functional accesses to the gem5/SST bridge to
boot disk images using the bridge. The change also adds more
information in INSTALL.md for MacOS users on how to get the bridge
setup working.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext: added functional accesses to gem5/SST bridge.

This change adds functional accesses to the gem5/SST bridge to
boot disk images using the bridge. The change also adds more
information in INSTALL.md for MacOS users on how to get the bridge
setup working.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
Merge branch 'kg/dm-numa-sst' of https://github.com/darchr/gem5 into kg/dm-numa-sst

powerjg commented on ext/sst/translator.hh:
```diff
@@ -144,27 +220,45 @@ gem5RequestToSSTRequest(gem5::PacketPtr pkt,
         sst_request_id_to_packet_map[request->getID()] = pkt;
-
+    
+    // if(gem5::curTick() > 340330000000)
+        // std::cout << request->getString() << std::endl;
```
Don't include commented out code.

kaustav-goswami replied:
fixed it in the latest patchset.

powerjg commented on src/sst/outgoing_request_bridge.cc:
```diff
@@ -105,19 +107,48 @@ OutgoingRequestBridge::sendTimingSnoopReq(gem5::PacketPtr pkt)
 OutgoingRequestBridge::
 OutgoingRequestPort::recvAtomic(PacketPtr pkt)
 {
+    // return 0;
```
Remove this

kaustav-goswami replied:
done

powerjg commented on src/sst/outgoing_request_bridge.cc:
```diff
@@ -105,19 +107,48 @@ OutgoingRequestBridge::sendTimingSnoopReq(gem5::PacketPtr pkt)
-    initData.push_back(std::make_pair(pkt->getAddr(), data));
+    // This should not receive any functional accesses
+    gem5::MemCmd::Command pktCmd = (gem5::MemCmd::Command)pkt->cmd.toInt();
+    // std::cout << "Recv Functional : 0x" << std::hex << pkt->getAddr() <<
+    // std::dec << " " << pktCmd << " " << gem5::MemCmd::WriteReq << " " <<
+    // getInitPhaseStatus() << std::endl;
```
Remove this

kaustav-goswami replied:
done

powerjg commented on src/sst/outgoing_request_bridge.cc:
```diff
@@ -84,6 +85,7 @@ OutgoingRequestBridge::getAddrRanges() const
 std::vector<std::pair<Addr, std::vector<uint8_t>>>
 OutgoingRequestBridge::getInitData() const
 {
+    std::cout << "getInitData() called!" << std::endl;
```
Remove this

kaustav-goswami replied:
done

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -241,6 +263,44 @@ SSTResponderSubComponent::handleRecvRespRetry()
+            break;
+        }
+        default:
+            panic("handleRecvFunctional: Unable to convert gem5 packet: %s\n", pkt->cmd.toString());
```
Line is >80 chars

kaustav-goswami replied:
should be fixed in the latest patchset

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -241,6 +263,44 @@ SSTResponderSubComponent::handleRecvRespRetry()
+    // SST does not understand what is a functional access in gem5. Since it
+    // has all the stored in it's memory, any functional access made to SST has
+    // to be correctly handled. All functional access *must be* writes.
+    std::cout << "handleRecvFunc was called! Need to do something here!" << std::endl;
```
Remove this

kaustav-goswami replied:
fixed it

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -108,6 +108,9 @@ SSTResponderSubComponent::init(unsigned phase)
         }
+        responseReceiver->initData.clear();
+        responseReceiver->initPhaseComplete(true);
+        std::cout << "Init phase is complete " << responseReceiver->getInitPhaseStatus() << std::endl;
```
Remove

kaustav-goswami replied:
done

powerjg commented on ext/sst/sst_responder.cc:
```diff
@@ -65,4 +65,6 @@ void
 {
     owner->handleRecvFunctional(pkt);
+    // std::cout << "SSTResponder" << std::endl;
+    // make these timing?
```
Remove

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -205,6 +208,25 @@ SSTResponderSubComponent::portEventHandler(
+        else if (SST::Interfaces::StandardMem::ReadResp* test =
+                dynamic_cast<SST::Interfaces::StandardMem::ReadResp*>(
+                request)) {
+            // functional calls this
+            // gem5::RequestPtr req = std::make_shared<gem5::Request>(
+            //     dynamic_cast<SST::Interfaces::StandardMem::ReadResp*>(
+            //         request)->pAddr,
+            //     dynamic_cast<SST::Interfaces::StandardMem::ReadResp*>(
+            //         request)->size, 0, 0);
+
+            // gem5::PacketPtr pkt = new gem5::Packet(
+            //     req, gem5::MemCmd::ReadResp);
+
+            // // Clear out bus delay notifications
+            // pkt->headerDelay = pkt->payloadDelay = 0;
+            // if (!(responseReceiver->sendTimingResp(pkt)))
+            //     responseQueue.push(pkt);
```
Why is all of this commented out code included?

kaustav-goswami replied:
this was done to test adding a new memory packet to gem5 on the fly. the issue is that it does not have a destination, so the simulation crashes. i have removed all this code now as it serves no purpose.

powerjg commented on src/sst/outgoing_request_bridge.hh:
```diff
@@ -97,8 +109,18 @@ class OutgoingRequestBridge: public SimObject
+
+    // We read the value of the init_phase_bool using `getInitPhaseStatus`
+    // method.
+
```
This should not be a newline... overall, the style is not up to the guidelines of gem5's style guide.

kaustav-goswami replied:
ive added a new newline. i'll update the file with smaller patches to adhere with the gem5's style.

powerjg commented on src/sst/outgoing_request_bridge.cc:
```diff
@@ -119,7 +118,7 @@ void
 {
     // This should not receive any functional accesses
-    gem5::MemCmd::Command pktCmd = (gem5::MemCmd::Command)pkt->cmd.toInt();
+    // gem5::MemCmd::Command pktCmd = (gem5::MemCmd::Command)pkt->cmd.toInt();
```
Don't leave commented out code.

kaustav-goswami replied:
fixed it in the latest patchset.

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -263,19 +263,22 @@ SSTResponderSubComponent::handleRecvRespRetry()
+    // only allows functional accesses at init time. Since it
     // has all the stored in it's memory, any functional access made to SST has
-    // to be correctly handled. All functional access *must be* writes.
+    // to be correctly handled. The idea here is to convert this timing access
```
```suggestion
to be correctly handled. The idea here is to convert this functional access
```

kaustav-goswami replied:
done

powerjg commented on ext/sst/Makefile:
Why is this file deleted in this commit?

powerjg commented on ext/sst/INSTALL.md:
````diff
@@ -98,6 +98,13 @@ According to the OS that you're using, you need to rename the `Makefile.xxx` to
+```sh
+# go to the base gem5 directory
+cd ../..
+export DYLD_LIBRARY_PATH=:`pwd`/build/RISCV/
````
Why specifically `build/RISCV/`?

This would be better if it was generic. E.g., `export DYLD_LIBRARY_PATH=:$GEM5_BASE/<build directory>` 

This being said, this change has *nothing* to do with "added functional accesses to gem5/SST bridge" it should be in a different change.

kaustav-goswami replied:
we are supporting the RISCV version as the default one. We can definitely do this but we'd also need to change the README and the INSTALL files to make it more generic. I agree on making it more generic as the default examples we have are `example.py` -> RISCV/default and `arm_example.py` -> ARM.

powerjg commented on ext/sst/gem5.hh:
```diff
@@ -141,11 +141,11 @@ class gem5Component: public SST::Component
-        {"system_port", "Connection to gem5 system_port", "gem5.gem5Bridge"},
-        {"cache_port", "Connection to gem5 CPU", "gem5.gem5Bridge"}
-    )
+    // SST_ELI_DOCUMENT_SUBCOMPONENT_SLOTS(
```
WHy this change? and why in this commit?

kaustav-goswami replied:
This was updated in this patchset. It spilled over to this change as I felt it was no longer necessary to explicitly document system_port and memory_port as we are now using a vector of outgoing ports

powerjg reviewed (changes requested):
This is a very quick review... can you please clean up the commits before sending this out for review? There are changes in commits that should be in other commits, and please don't use "merge" and instead use **rebase**. When there are merge commits in a PR, I cannot tell what are changes that I should be reviewing vs changes that are actually already in upstream.

kaustav-goswami committed:
ext: outgoing bridge supports limited func accesses

This change adds a lot of new features to the gem5-SST bridge.
The following list of changes describes each of them:
1. The bridge now supports limited functional accesses, which
enables users to work with vda devices in ARM/RISCV systems.
2. There are new methods added to the bridge which allows future
extension of the bridge to swap ports to enable delayed INIT
phase for SST. This will allow the bridge to start the system in
gem5 and then swap the memory ports to SST at RUN phase.
3. A lot of stray debug statements/cout statements were removed
in this patch.
4. New documentation was added to the SST component for gem5.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext: corrected minor changes in the outgoing bridge

This change fixes minor changes with the gem5's outgoing bridge,
mostly in comments and gem5 syle.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext,configs: drops and adds numa-enabled stdlib boards

This change drops all the boards and configs created previously
to simulate disaggregated memory systems as those were not strictly
adhering to the stdlib rules. This new change posts most of the
previously written code to create a new abstract disaggregated
memory board for both ARM and RISCV architectures, which minimizes
code duplication and can be easily extended. Extension suggestions
and tips are now commented thoughout the python files. This is a
proper directory structure to navigate these changes.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext,configs: added stdlib examples to sst scripts

This change create config script for both gem5 and SST and allows
SST to run these scripts using one or more ISAs at the same time.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami commented:
@powerjg I am sorry that I had such a messy code. I restructured the whole disaggregated memory code. I had a conversation with @studyztp who was trying to extend this code for her work and it was not trivial for her. In this version, I have split the whole setup into `boards`, `cachehierarchies`, `memories` and `configs`, which should be easy to read and extend. The code extends the stdlib in a systematic manner. I have also removed a lot of unnecessary config files for both gem5 and SST and made the new configs flexible.

kaustav-goswami committed:
configs: updated numactl calls

This change updates the numactl calls to correctly allocate memory
in the local and the remote memory nodes.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext,base-stats: add stats to the gem5 SST bridge

This change adds stats counters to the gem5 SST external bridge to
monitor the number and size of packets going out and coming into
gem5. This numbers should also be reflected in SST's stats output
file.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs,stdlib: added a private l1, l2 and shared l3 cache

This change adds a private l1 private l2 and shared l3 cache
hierarchy as a stdlib component.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs: updated disaggregated memory scripts

This change updates the disaggregated memory scripts to use a 3-level
cache and dumps starts after each STREAM kernel.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext,base-stats: minor fix in the stat counter

This commit fixes a compilation error on the gem5 SST bridge.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext,base-stats: minor fix in the stat counter

This commit fixes a compilation error on the gem5 SST bridge.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext: fixed repeated lines of code

This patch fixes the previous git merge that duplicated a few lines
in the ext library of sst.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
Merge branch 'kg/dm-numa-sst' of github.com:darchr/gem5 into kg/dm-numa-sst

kaustav-goswami committed:
ext: fixed ext: fixed repeated lines of code

This commit further fixes repeted lines of code.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
Merge branch 'kg/dm-numa-sst' of github.com:darchr/gem5 into kg/dm-numa-sst

powerjg commented on disaggregated_memory/boards/arm_sst_board.py:
```diff
@@ -0,0 +1,239 @@
+
+    @overrides(ArmAbstractDMBoard)
+    def _set_remote_memory_ranges(self):
+        pass
```
Error?

powerjg replied:
Should you add an error here?

powerjg commented on disaggregated_memory/boards/arm_sst_board.py:
```diff
@@ -0,0 +1,239 @@
+            clk_freq=clk_freq,
+            processor=processor,
+            local_memory=local_memory,
+            remote_memory_addr_range=self._remoteMemory.remote_memory.physical_address_ranges[
```
This should probably be a function of the remote memory.

powerjg commented on disaggregated_memory/boards/arm_sst_board.py:
```diff
@@ -0,0 +1,239 @@
+        # we need to find whether there is any external latency. if yes, then
+        # add xbar to add this latency.
+
+        if self.get_remote_memory().is_xbar_required():
```
We should try to make this more generic

powerjg commented on disaggregated_memory/boards/riscv_dm_board.py:
```diff
@@ -0,0 +1,482 @@
+        )
+
+    @overrides(RiscvBoard)
+    def generate_device_tree(self, outdir: str) -> None:
```
How is this different from the base?

powerjg commented on disaggregated_memory/boards/riscv_sst_board.py:
```diff
@@ -0,0 +1,244 @@
+
+        :returns: The memory system.
+        """
+        raise NotImplementedError
```
?? This is odd. Can you explain why this is not implemented?

powerjg commented on disaggregated_memory/boards/riscv_sst_board.py:
```diff
@@ -0,0 +1,244 @@
+        # we need to find whether there is any external latency. if yes, then
+        # add xbar to add this latency.
+
+        if self.get_remote_memory().is_xbar_required():
```
This xbar should bein the remote memory object and simply exposed as a single port. 

Let's generalize this by assuming the remote memory will always be a single port (i.e., never channeled). Then, you can make this function be exactly the same for sst/gem5 and arm/riscv.

In fact, there should be a "mixin" class that allows disaggregated memory.

powerjg commented on disaggregated_memory/cachehierarchies/dm_caches_sst.py:
```diff
@@ -0,0 +1,218 @@
+# Copyright (c) 2023 The Regents of the University of California
```
What are "dm" caches and why are they different from the normal cache hiearchies?

powerjg commented on disaggregated_memory/memories/remote_memory.py:
```diff
@@ -0,0 +1,92 @@
+        ]
+
+    @overrides(ChanneledMemory)
+    def get_mem_ports(self) -> Sequence[Tuple[AddrRange, Port]]:
```
Let's make this a single port and put the xbar in this object.

powerjg commented on disaggregated_memory/memories/external_remote_memory.py:
```diff
@@ -0,0 +1,86 @@
+from m5.objects import OutgoingRequestBridge, AddrRange, Tick
+
+
+class ExternalRemoteMemoryInterface:
```
I think this should inherit from `AbstractMemorySystem`

powerjg commented on disaggregated_memory/configs/riscv-sst-numa-nodes.py:
```diff
@@ -0,0 +1,164 @@
+board._pre_instantiate()
+root = Root(full_system=True, board=board)
+board._post_instantiate()
+m5.instantiate()
```
These four lines are very odd. Can you explain what it's doing?

kaustav-goswami committed:
ext: updated the example arm dm script

This change makes the remote memory a 4 channel DDR4-like memory
installed on the remote node.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext: restrict any functional reads to go to sst

This change restricts any functional reads at runtime to go to
SST.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs: added a KVM capable ARM NUMA runscript

This change adds a KVM capable runscript for testing the NUMA
setup.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
arch-x86,configs: added a WIP x86 dm board

This change adds a WIP working x86 board that allows memory
offlining with compatible kernel.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs: added a arm kvm config for dm

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>
